crates/daphne-worker/src/durable/replay_checker.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2025 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+use super::{req_parse, GcDurableObject};
+use crate::int_err;
+use daphne::messages::{AggregationJobId, ReportId};
+use daphne_service_utils::durable_requests::bindings::{self, replay_checker, DurableMethod};
+use std::{
+    collections::{HashMap, HashSet},
+    iter::zip,
+    sync::OnceLock,
+    time::Duration,
+};
+use wasm_bindgen::JsValue;
+use worker::{js_sys, Env, Request, Response, Result, ScheduledTime, State};
+
+super::mk_durable_object! {
+    /// Where report ids are stored for replay protection.
+    struct ReplayChecker {
+        state: State,
+        env: Env,
+        seen: HashMap<ReportId, AggregationJobId>,
+    }
+}
+
+impl GcDurableObject for ReplayChecker {
+    type DurableMethod = bindings::AggregateStore;
+
+    fn with_state_and_env(state: State, env: Env) -> Self {
+        Self {
+            state,
+            env,
+            seen: Default::default(),
+        }
+    }
+
+    async fn handle(&mut self, mut req: Request) -> Result<Response> {
+        match replay_checker::Command::try_from_uri(&req.path()) {
+            Some(replay_checker::Command::Check) => {
+                let replay_checker::Request {
+                    report_ids,
+                    aggregation_job_id,
+                } = req_parse(&mut req).await?;
+
+                let mut duplicates = HashSet::new();
+
+                let report_ids_as_string = report_ids
+                    .iter()
+                    .filter(|r| match self.seen.get(r) {
+                        Some(cached_agg_job_id) => {
+                            if *cached_agg_job_id != aggregation_job_id {
+                                duplicates.insert(**r);
+                            }
+                            false // skip checking
+                        }
+                        None => true, // check against disk
+                    })
```

Comment:
Can you add a comment explaining what's going on here? I'm having trouble parsing this.

FWIW: We need to disallow replays across aggregation jobs, so I'm not sure what role the agg job ID plays here.

Reply:
Documented the DO itself